### PR TITLE
Support for StrictMath.log1p

### PR DESCRIPTION
**Pull request by AnishChandran**

### CODE REVIEW

**Comment by jvilk** [comment]
It looks like `Math.log1p` is not implemented in all browsers (see [MDN for compatibility details](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/log1p)). We will need a polyfill for browsers that do not support it.

Could you add a polyfill to `src/util.ts`, in [a similar way that we added `Math.imul`]?

Thanks!

File: classes/test/StrictMathTest.java [hunk 1]
```diff
@@ -261,15 +268,6 @@ public static void main(String[] args) {
     }
     results("double pow(double a, double b)", d_results_2d);
 
-    // static double hypot(double a, double b)
-    // Returns the value of the first argument raised to the power of 2 + second argument raised to the power of .
-    for (i=0; i < d_vals.length; i++) {
```
Comment (perimosocordiae) [inline classes/test/StrictMathTest.java]: Why is this getting deleted?

File: src/natives/java_lang.ts [hunk 2]
```diff
@@ -919,9 +919,7 @@ class java_lang_StrictMath {
   }
 
   public static 'log1p(D)D'(thread: threading.JVMThread, d_val: number): number {
-    thread.throwNewException('Ljava/lang/UnsatisfiedLinkError;', 'Native method not implemented.');
-    // Satisfy TypeScript return type.
-    return 0;
+    return Math.log1p(1);
```
Comment (perimosocordiae) [inline src/natives/java_lang.ts]: Seems like this should be `Math.log1p(d_val);`

File: classes/test/StrictMathTest.java (hunk 1 above)
Comment (AnishChandran) [inline classes/test/StrictMathTest.java]: Actually I have no idea how this happened ! sorry ! is the rest good ?

**Comment by perimosocordiae** [comment]
It looks like you made this PR off an out-of-date local workspace. Please make sure to only commit the changes you intended to make.

**Comment by AnishChandran** [comment]
Sorry ! I just kept it Math.log1p(1) to check the values and forgot to change it back ! I am extremely sorry for the inconvenience caused !

**Comment by AnishChandran** [comment]
I think It may cause problems because the input for the log functions should be >=1 but we are giving negative values here! That may be the cause of problems ?

**Comment by jvilk** [comment]
Your function should be able to appropriately handle any valid numerical value entering the function, including edge cases like `Infinity` and `NaN`. Its behavior should match that of the native HotSpot JVM.

**Comment by perimosocordiae** [comment]
A useful strategy here is to run the tests once, then check the output from native java in the generated StrictMath.runout file. As long as you match that, you'll have the right behavior for the edge cases we care about.

**Comment by AnishChandran** [comment]
Is there a way to include the isNan()/Number.Nan static function in
typescript ?
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/NaN

I had a look at the
https://typescript.codeplex.com/workitem/1085

Is there any possible way

**Comment by perimosocordiae** [comment]
An easy workaround for isNan is to check `x === x`. It will only be false when `x` is NaN.

**Comment by jvilk** [comment]
Huh? `isNaN` should be defined. Use that function.

https://github.com/plasma-umass/doppio/blob/927371e330b3a759b4eb7157df6102fc38df2ff3/src/gLong.ts#L97

**Comment by jvilk** [comment]
And it's `isNaN`, not `isNan`.

File: src/natives/java_lang.ts [hunk 3]
```diff
@@ -919,9 +919,14 @@ class java_lang_StrictMath {
   }
 
   public static 'log1p(D)D'(thread: threading.JVMThread, d_val: number): number {
-    thread.throwNewException('Ljava/lang/UnsatisfiedLinkError;', 'Native method not implemented.');
-    // Satisfy TypeScript return type.
-    return 0;
+    if(d_val < -1 || Number.isNaN(d_val))
```
Comment (jvilk) [inline src/natives/java_lang.ts]: Use `isNaN`. Older browsers don't define `isNaN` on `Number`, and all browsers define it globally.

Comment (jvilk) [inline src/natives/java_lang.ts]: Also, this is incorrect. `StrictMath.log1p(-1)` is `-Infinity`, not `NaN`. Did you run the unit test? It should fail on -1.

In addition, `Math.log1p

File: src/natives/java_lang.ts [hunk 4]
```diff
@@ -919,9 +919,14 @@ class java_lang_StrictMath {
   }
 
   public static 'log1p(D)D'(thread: threading.JVMThread, d_val: number): number {
-    thread.throwNewException('Ljava/lang/UnsatisfiedLinkError;', 'Native method not implemented.');
-    // Satisfy TypeScript return type.
-    return 0;
+    if(d_val < -1 || isNaN(d_val))
+      return Number.NaN;
+    else if(d_val == -1)
```
Comment (AnishChandran) [inline src/natives/java_lang.ts]: I have included the value for log1p(-1) as negativeinfinity here !

**Comment by AnishChandran** [comment]
I dunno what i s wrong ! it is not executing crctly for Double.MIN_VALUE and Double.MIN_NORMAL ! i cant find anything wrong in the code too ! i have tried all possibilities from http://pastebin.com/MTFFGktZ to just return Math.log(1+val)

can you help me out ?

**Comment by AnishChandran** [comment]
http://pastebin.com/5H3iPBVQ these are the output details ! kindly take a look once whenever possible !
